drivers/hcsr04/hcsr04.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2021 AUTH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_hcsr04
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the HC-SR04
+ *
+ * @author      Evripidis Chondromatidis <eurichon1996@gmail.com>
+ *
+ * @}
+ */
+
+#include "hcsr04.h"
+#include "hcsr04_constants.h"
+#include "hcsr04_params.h"
+
+#include "periph/gpio.h"
+#include "xtimer.h"
+
+#include <stdint.h>
+#include <errno.h>
+#include <math.h>
+
+
+void hcsr04_int_callback(void *arg){
+    uint32_t now = xtimer_now_usec();
+    hcsr04_t *p_dev = (hcsr04_t *)arg;
+
+    if (gpio_read(p_dev->params.echo_pin)) {
+        p_dev->pre_meas_t_us = now;
+    } else {
+        uint32_t dt_us_one_way = (now - p_dev->pre_meas_t_us) / 2;
+        p_dev->distance = 1e-3 * p_dev->sound_speed * dt_us_one_way;
+    }
+}
+
+int hcsr04_init(hcsr04_t *dev, const hcsr04_params_t *params)
+{
+    if (!dev || ! params) {
+        return -EINVAL;
+    }
+
+    dev->params = *params;
+
+    dev->distance = 0;
+    dev->pre_trig_t_us = xtimer_now_usec();
+    dev->pre_meas_t_us = xtimer_now_usec();
+    hcsr04_set_temp(dev, (*params).temperature);
+
+    if (gpio_init(dev->params.trigger_pin, GPIO_OUT)) {
+        return -EIO;
+    }
+
+    if (gpio_init_int(dev->params.echo_pin, GPIO_IN, GPIO_BOTH, hcsr04_int_callback, (void *)(dev))) {
+        return -EIO;
+    }
+
+    xtimer_msleep(30);
+
+    uint16_t distance;
+    int status = hcsr04_get_distance(dev, &distance);
+
+    if (dev->distance == 0) {
+        return -ECON;
+    }
```

Comment:
When a read is triggered the driver waits a certain time before the interrupt gets called and the `dev->distance` value gets updated right? I think it could be better to use a `mutex_t` on the device descriptor `hcsr04_t` for this.

Initially you lock the mutex before triggering the read, e.g.:
1. Lock the mutex (`mutex_lock(&dev->lock)`)
2. Then trigger the read.
3. Try to lock it again (`mutex_lock(&dev->lock)`) or with (`xtimer_mutex_lock_timeout(&dev->lock, min_wait)`) if the device may not respond and check for `distance == 0`
4. On the IRQ callback unlock the mutex to continue the execution on point 3.

This way the value should be always synchronized.